src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/ImmutableArrayExtensions.cs

```diff
@@ -84,4 +85,20 @@ static ImmutableArray<TResult> BuildResult(ImmutableArray<T> items, Func<T, TRes
             return result.DrainToImmutable();
         }
     }
+
+    public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IReadOnlyCollection<TSource>? source, Func<TSource, IEnumerable<TResult>> selector)
```

Comment:
```suggestion
    public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IReadOnlyCollection<TSource>? source, Func<TSource, ImmutableArray<TResult>> selector)
```

We're not using it more generally at this point, so we can avoid a box. If a more general version is needed later, we can do that then.

Reply:
> we can avoid a box

Wouldn't be so sure about that - the called `AddRange` method has only `IReadOnlyList` or `IEnumerable` overloads (and even the former one calls into `IEnumerable` one inside). Added more `ImmutableArray` overloads down the chain.